HFE.md

```diff
@@ -0,0 +1,643 @@
+
+
+### Alternatives considered
+
+Another potential solution would be to always place the expiry at the **END** of the entry allocation:
+
+![Alt text](images/HFE-images/image5.png)
+
+This has the benefit of avoid reallocating EXISTING entries when an expiry is first set or removed (e.g. using HEXPIRE or HPERSIST). The reason no choosing this potential layout is that it might lead to cases of cache misses when accessing expiry data for entries with long field/value.     
+
+
+#### HASH Object ListPack representation
+
+User can configure hash type objects to use memory efficient representation using listpacks. Although we could add TTL to the listpack representation this might greatly complicate the usage when some items do not have TTL and others does. Using an external data structure to manage the TTL for listpack elements which were assigned TTL is possible but might reduce the memeory efficiency of this representation. In the sake of simplicity we will force listpack conversion to hashtable representation as soon as the first item TTL is set. This way we will not enforce memory efficiency degradation for users not setting TTL on hash object elements.   
+
+
+### Items expiration context
+
+When expiring an entry, the following actions will take place:
+
+1. Propagate the `HDEL` event to the replication backlog
+2. remove the relevant entry from the containing Object (HASH, SET, ZSET etc...)
+3. Untrack the entry for active expiration (will be extended later)
+4. issue a keyspace notification regarding the item expiry
+5. update expiry related statistics
+
+In order to support all these operations there are mandatory inputs which must be provided for each expiration action:
+
+1. The current database - is required in order to propagate the command in the correct database context.
+2. The relevant Object - is required for several reasons. 
+    **First** since all objects have their key name embedded we can retain the key name in order to propagate the command to the replication backlog. **
+    Second** all related item tracking can be accessed and updated from the object itself.
+
+### Lazy expiration
+
+Lazy expiration happen whenever an item which has expired TTL is accessed. In this case the item will be **immediately** removed from the owning object and expired (as described in the expiration context section). This include both read and write commands access context. For items (fields) ‘access’ logic we aim to follow the same logic used for generic keys. 
+When accessing an item, there are 3 potential states it can be stated:
+
+1. VALID - meaning the item is to be treated as any existing item and be included in replies as well as operated on during actions and mutations.     
+2. INVALID - meaning the item should be treated as “not exists” and is NOT to be included in any operations and/or replies.
+3. INVALIDATED(deleted) - meaning the item is to-be-removed immediately (i.e expired) and thus does not exist anymore. 
```

Comment:
The terms **VALID**, **INVALID**, and **INVALIDATED** are too similar and not intuitive. INVALID suggests corruption, but an expired field is still well-formed and valid. Consider clearer terms like LIVE/NORMAL, EXPIRED, and DELETED/PURGED to reflect TTL state accurately.

Reply:
I agree. In general, I think we should use the word "expired" to mean that the time has passed for a key, whether it still exists in memory or not. An expired key can exist in the database if we haven't deleted it yet.

We should not use this word as a transitive verb as in "to expire a key". Instead we should say "to delete an expired key".

Reply:
I agree about this terminology in the scope of lazy expiration (i.e. in the RFC). I used these terms since these are the states recognized by the hashtable which I would like to keep out of scope of  "expiration" and "volatile".
I will change it in the RFC.

Reply:
@xbasel / @zuiderkwast  fixed. looks more to your point?